### ISSUE TITLE
Move stylesheet editor from webforms to Angular.

### ISSUE THREAD

**Issue by madsrasmussen** [umbraco/Umbraco-CMS#3438]
Functionality should stay the same.

**Comment by kjac** [comment]
I'll give this a shot.

**Comment by kjac** [comment]
@madsrasmussen so I've been thinking. Is the current functionality with the RTE styles as sub nodes to the stylesheet really the best, "don't-make-me-think" UX? 

I don't think it is 😄 you really have to know the feature is there to use it.

The way I see it, the CSS editor and the property mapping for the RTE should be in the same editor in a tabbed UI - sort of like what we have for doctypes:

![image](https://user-images.githubusercontent.com/7405322/47678402-03482c00-dbc2-11e8-9ab7-1c02ad7dd684.png)

This will have a few benefits:

1. The UI becomes less secretive - no "hidden" features.
2. No need for additional CSS rules defined in a text area on the sub nodes - the CSS only exists within the editor.
3. No need to generate the CSS from sub nodes, which will save a bunch of complexity and remove the "umb_name: xxx" annotations within the CSS.

Thoughts?

**Comment by dawoe** [comment]
Maybe we should consider storing custom styles used in the editor dropdown another way.

I always use this way using tinyMce config https://www.justnik.me/blog/umbraco-rte-styles-another-way

This avoids unwanted span-tags being inserted etc...

Dave

**Comment by kjac** [comment]
@dawoe that'd definitively make the complexity of stylesheet editing even lower. But configuration approach aside, won't this way of setting up the RTE styles limit us to always use the same styles for all RTE's? As it is now you can pick a stylesheet for an RTE datatype and those styles will be available in the RTW.

**Comment by kjac** [comment]
We need some HQ decision makers here 😁

**Comment by madsrasmussen** [comment]
Hello gentlemen,

I agree with your points about the current stylesheet editor. It is too difficult to figure out how to configure styles for the RTE.

For v8 I think it makes sense to move the RTE subnodes to a "tab" but I can't tell from your picture how it is going to work. Is it possible to do a screenshot, video, or finish the PR (depending on how you work the best :) ) so we can see the full editing flow?

The long-term solution might be something entirely different. We haven't found a good solution yet to bring styles and RTE config closer together but that will be after v8.

The most important part of this task is to get the stylesheet editor working in v8. We also need to make sure the output of the stylesheets are the same for the new editor, so the upgrade path is as smooth as possible.

**Comment by dawoe** [comment]
Just as a reference for the future. The custom style formats can also be set at runtime instead of through config 
https://www.tiny.cloud/docs/demo/format-custom/

**Comment by kjac** [comment]
@madsrasmussen what I'm proposing is a "Styles"-tab where you edit metadata about the CSS - e.g. "When you select 'Heading' in the RTE style selector, apply the class '.heading'". But that metadata would be stored in the DB - as opposed to the current solution, where the style sheet sub nodes are actually extracted from (and written to) the CSS file by means of the "umb_name" magic.

But it won't be backwards compatible. Perhaps a migration could be created to mediate this, but that's way out of scope for the time I have at hand.

In other words... if backwards compatibility is a requirement, I'll go ahead and re-implement the current functionality 😄

**Comment by dawoe** [comment]
Actually the style definitions were stored in a the database in Umbraco pre 7.3.0

If you look in the source code :
 https://github.com/umbraco/Umbraco-CMS/blob/dev-v7/src/Umbraco.Core/Persistence/Migrations/Upgrades/TargetVersionSevenThreeZero/MigrateStylesheetDataToFile.cs
and
https://github.com/umbraco/Umbraco-CMS/blob/dev-v7/src/Umbraco.Core/Persistence/Migrations/Upgrades/TargetVersionSevenThreeZero/RemoveStylesheetDataAndTables.cs

There were probably good reasons to move it away from the database. But can't figure that out from the issue on the old tracker : https://issues.umbraco.org/issue/U4-114

**Comment by Shazwazza** [comment]
CSS hasn't been stored in the database for years and nor should it. The 'special' styles were achieved with the `umb_name` parsing magic which seems to work fine. IMO if we can 'just' move the sub-styles editor into a dialog/panel or whatever instead of having each style as sub-nodes in the tree that would be heaps better and we can just continue to use the same file formatting as v7 since it's not causing any harm and works ok for now.

**Comment by kjac** [comment]
@Shazwazza I absolutely agree. If only it was so simple 🤔 

Problem is the two views both edit the same data:

- The CSS editor obviously edits the entire CSS file.
- The property editor edits chunks of the same CSS file, which is then stitched together on the server side by means of `umb_name` magic.

If there is no page load between the two editors (e.g. in tabs on the same page as suggested earlier), they get out of sync really really fast. We then need to duplicate the `umb_name` magic handling on the client side, and that's a big no-go for me.

I'll investigate the options for a server callback when switching between tabs. That would solve the problem, even if it'd put a bit more load on the server. Luckily this isn't a high traffic area of the back office 😄

**Comment by bjarnef** [comment]
@kjac looks great so far. Maybe adding styles should be handled from a modal/dialog similar to when adding new properties on document types? Otherwise I think the sorting will be difficult the use (unless they are collapsed when in sorting mode like on document types).

It would also give a nicer overview of the styles, when these are collapsed.

**Comment by kjac** [comment]
@bjarnef excellent idea! See #3503 for at demo GIF 😄

**Comment by nul800sebastiaan** [comment]
Forgot to close this one :)
